### ISSUE TITLE
What about using LSP4IJ?

### ISSUE THREAD

**Issue by angelozerr** [clojure-lsp/clojure-lsp-intellij#53]
I noticed that you have implemented your own lsp support, but I would like to notify you that it exists now a free LSP support for IntelliJ https://github.com/redhat-developer/lsp4ij

 * You can see the current LSP support at https://github.com/redhat-developer/lsp4ij/blob/main/docs%2FLSPSupport.md
By reading your LSP support at https://github.com/clojure-lsp/clojure-lsp-intellij/blob/master/docs/capabilities.md it seems LSP4IJ covers the same thing than your plugin and perhaps more. I have tested quickly https://github.com/clojure-lsp/clojure-lsp-intellij/issues/51 and it seems it is working with LSP4IJ.

 * LSP4IJ provides an LSP console to show LSP traces and see the state of language server https://github.com/redhat-developer/lsp4ij/blob/main/docs%2FUserGuide.md

 * If you want to play quickly with LSP4IJ and your Clojure LS without developping an IJ plugin please read https://github.com/redhat-developer/lsp4ij/blob/main/docs/user-defined-ls/clojure-lsp.md

 * If you want to integrate your ls in an IJ plugin please read https://github.com/redhat-developer/lsp4ij/blob/main/docs%2FDeveloperGuide.md

You can read an article about LSP4IJ at https://idetools.dev/blog/lsp4ij-announcement/

Hope you could be interested with LSP4IJ and don't hesitate to answer me if you have any questions.

Thanks!

**Comment by ericdallo** [comment]
Hey @angelozerr, thanks for bringing that, I was not aware of that and it's awesome to see free open source alternatives like that one! I'll take a closer look, but some things I noticed:

- When I hold `Ctrl`, and hover over a symbol, I see the whole buffer text underlined, seems a little bit weird
- Do you support semantic tokens?
- When I click on a code lens I see this, which I'm not entirely sure what means:
![image](https://github.com/clojure-lsp/clojure-lsp-intellij/assets/7820865/ae86420e-64ba-4b01-ab7e-f98dcf25d97e)

When I created this plugin, the idea is to provide LSP support but in a way intellij users don't need to setup nothing, like just works out of the box, with minor customization as possible, is it possible to have a plugin that uses lsp4ij under the hood and have all those things configured already out of the box?

**Comment by angelozerr** [comment]
> When I hold Ctrl, and hover over a symbol, I see the whole buffer text underlined, seems a little bit weird

I think it is the same issue than https://github.com/redhat-developer/lsp4ij/issues/98 It is because if you have use just TextMate or TEXT language, the PsiFile is not tokenized and the range is the full file content. It is badly not possible to customize that with gotoDeclarationHandler extension point -( But if you have a plugin which tokenized the PsiFile with PsiElement you will not have this problem.

> Do you support semantic tokens?

I have pushed last hours this support, see doc at https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserGuide.md#semantic-tokens-support but it is experimental and it seems that there is a CPU issue https://github.com/redhat-developer/lsp4ij/issues/394 but if you want to play with this feature you can install nightly build https://github.com/redhat-developer/lsp4ij?tab=readme-ov-file#testing-nightly-builds

I have played with clojure lsp and it seems it colorizes some tokens.

> When I click on a code lens I see this, which I'm not entirely sure what means:

It means that language server report Codelens with command `code-lens-references`, you can see those CodeLens by using LSP console https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserGuide.md#lsp-console

If language server doesn't provide this command, it tries to search if it exists a command on client side, if this command doesn't exist, you have this popup error and you have a link which suggest to implement an IJ plugin which register the expected command.

I think you have implemente this LSP command in your plugin, right?

> clojure-lsp-intellij has a good feedback during initialization which I don't see easily, I need to click on LanguageServers and even so have logs enabled, check the $/progress notifications

$/progress has been implemented with IntelliJ Backgroind Task, see https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md#progress-support

What is your suggestion to improve that?

> When I created this plugin, the idea is to provide LSP support but in a way intellij users don't need to setup nothing, like just works out of the box, with minor customization as possible, is it possible to have a plugin that uses lsp4ij under the hood and have all those things configured already out of the box?

See the section of our article https://idetools.dev/blog/lsp4ij-announcement/#how-to-integrate-your-language-server-in-an-intellij-plugin

It seems that you have played with https://github.com/redhat-developer/lsp4ij/blob/main/docs/user-defined-ls/clojure-lsp.md which is working just with few settings and you don't need to write an IJ plugin, but it is better to develop an IJ plugin:

 * to embed the clojure-lsp in the plugin
 * provide required LSP command like the `code-lens-references` command.

In other words, my suggestion is to update your IJ plugin, by removing all LSP features and  based on LSP4IJ and register the Clojure LSP with the LSP4IJ server extension point https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md

The goal of your plugin is to:

 * embed default clojure lsp
 * provide an action to download last version of clojure lsp
 * implement required LSP command like `code-lens-references` command

**Comment by ericdallo** [comment]
@angelozerr thanks for the explanation, it completly makes sense!
I wish we had that plugin when I started developing this one hehe, but I agree it makes sense to leave the LSP logic to LSP4IJ indeed, especially because of semantic tokens which I know is hard to support on client side, it's just that it's a huge transition we'd need to make carefully, especially because there are lots of people already using clojure-lsp-intellij plugin.

> But if you have a plugin which tokenized the PsiFile with PsiElement you will not have this problem.

clojure-lsp-intellij already has a language parser for Clojure, why that still happens?

So, just summarizing some things I have in mind so we can be aligned:
- I still need to test each existing feature and check it's working or the most important ones are working
- we would need to launch a 3.0.0 (next major version) of clojure-lsp-intellij, which would delegate all LSP connection to LSP4IJ
- I wanna a way to download clojure-lsp automatically the same way [it does ATM](https://github.com/clojure-lsp/clojure-lsp-intellij/blob/dc39be4f01d5c7b9694eccb2cb580c2600a4e4a9/src/main/clojure/com/github/clojure_lsp/intellij/server.clj#L139-L169), is there any way to have this logic working and after downloaded manually trigger the LSP start like we do today?
- about the code-lens-reference, got it, it should be easy to support it.
- we would need to fix this underlined issue
- Also, there is a behavior I'd like you to consider supporting, which is when Ctrl+click on a definition, we show the references, it's something I brought from Cursive which Intellij users use a lot instead the need to manually triggering the find-references shortcut, WDYT?

**Comment by ericdallo** [comment]
BTW, really cool project the LSP4IJ, it remembers me when we extracted clojure-lsp logic to https://github.com/clojure-lsp/lsp4clj 😄

**Comment by angelozerr** [comment]
> clojure-lsp-intellij already has a language parser for Clojure, why that still happens?

How have you tested LSP4IJ,with your plugin? If you did that you should have duplicate completion (one from your pluin, and one from LSP4IJ), right?

> I still need to test each existing feature and check it's working or the most important ones are working

Exactly, when I discovered your project, I have spend some time to check the feature that you have implemented before annoying you and I think a feature that you will loose is the code action available from the refactor menu. I have not implemented this feature. Perhaps LSP4IJ will need to provide some customization, I don't know, I'm waiting for feedback from adaptor like you if you decide to migrate to LSP4IJ.

> we would need to launch a 3.0.0 (next major version) of clojure-lsp-intellij, which would delegate all LSP connection to LSP4IJ

Indeed it will be more careful if you decide to adopt LSP4IJ.

> I wanna a way to download clojure-lsp automatically the same way [it does ATM](https://github.com/clojure-lsp/clojure-lsp-intellij/blob/dc39be4f01d5c7b9694eccb2cb580c2600a4e4a9/src/main/clojure/com/github/clojure_lsp/intellij/server.clj#L139-L169), is there any way to have this logic working and after downloaded manually trigger the LSP start like we do today?

Yes, LSP4IJ provide an API to enable/disable and start/stop language server. See https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#language-server-manager

> about the code-lens-reference, got it, it should be easy to support it.

I think.Please note that it exists the default command https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#editoractionshowreferences perhaps you could declare you action by using the class of this action?

> we would need to fix this underlined issue


> Also, there is a behavior I'd like you to consider supporting, which is when Ctrl+click on a definition, we show the references, it's something I brought from Cursive which Intellij users use a lot instead the need to manually triggering the find-references shortcut, WDYT?

Today Ctrl+Click is associate to the textDocument/definition, you woule like to associate to textDocument/references? I would like to avoid doing that because references can be expensive, but if you need it we could provide a customization or perhaps you could implement on your side the GoToDeclarationHandler with references. The main thing to do is to create issues on LSP4IJ to discuss about that

**Comment by ericdallo** [comment]
> How have you tested LSP4IJ,with your plugin? If you did that you should have duplicate completion (one from your pluin, and one from LSP4IJ), right?

I disabled my plugin, enabled yours and restarted intellij

> a feature that you will loose is the code action available from the refactor menu.

TBH, That's not a core feature, if code actions are provided via the context menu (lightbulb), that's its good enough. It'd be nice though to support that in the future

> provide an API to enable/disable and start/stop language server.

awesome, so I think that will keep working as expected

> perhaps you could declare you action by using the class of this action?

Ah, yeah, that should be enough!

> Today Ctrl+Click is associate to the textDocument/definition, you woule like to associate to textDocument/references?

Yeah, first time I heard about that I disliked too, but I understood later, the point is that if you are already on the definition, you don't want to go to definition, but the references of it, so users can use the same shortcut to go and come back.
IMO would be nice to have this in the core without the need to support any customization, but if you disagree, a customization would be really useful.

I'll find some time during the next week to deep dive and check if there are other bugs, thanks again for raising this.

**Comment by angelozerr** [comment]
> I disabled my plugin, enabled yours and restarted intellij

Ok so it means that you have not a PsiFile for clojure, right?

> TBH, That's not a core feature, if code actions are provided via the context menu (lightbulb), that's its good enough. 

It is and with fast performance (it should)

> It'd be nice though to support that in the future

Please create a new ossue for that with your detailled idea (ex : do you want some customization of the icon,etc)

> IMO would be nice to have this in the core without the need to support any customization, but if you disagree, a customization would be really useful.

Let's create a new issue to discuss about that and see the feedback from the community.

> I'll find some time during the next week to deep dive and check if there are other bugs, thanks again for raising this.

Great!

**Comment by ericdallo** [comment]
> Ok so it means that you have not a PsiFile for clojure, right?

Ah makes sense 😅, so if we had clojure-lsp-intellij using lsp4ij and defining a Parser like it does today, that would work, right? BTW, with lsp4ij, would be required to change anything related to Langauge declaration, like parser, lexer etc?

> Please create a new ossue for that with your detailled idea (ex : do you want some customization of the icon,etc)

Will do!

**Comment by angelozerr** [comment]
> > Ok so it means that you have not a PsiFile for clojure, right?
> 
> Ah makes sense 😅, so if we had clojure-lsp-intellij using lsp4ij and defining a Parser like it does today, that would worinspecter?

Yes!

BTW, with lsp4ij, would be required to change anything related to Langauge declaration, like parser, lexer etc?
> 

No LSP4IJ works with or without a PsiFile

> > Please create a new ossue for that with your detailled idea (ex : do you want some customization of the icon,etc)
> 
> Will do!
>

**Comment by angelozerr** [comment]
@ericdallo just for your information, we will release 0.6.0 next week and we provide a cool new feature with document symbol.

Here a screenshot with clojure:

![image](https://github.com/user-attachments/assets/4f490f6d-424d-4b7b-9698-5ac040edcbb8)

**Comment by ericdallo** [comment]
Awesome!
I still need to check the feature parity with [current features](https://github.com/clojure-lsp/clojure-lsp-intellij/blob/master/docs/features.md) so we can consider switch to LSP4IJ

**Comment by angelozerr** [comment]
> Awesome!

Glad it pleases you!

For some shortcut, I think it will not the same thing than you, since I have not found a solution to override the Action menu for Find implementation for instances https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md#implementation

I had to provide new shortcut. I don'tunderstand how you have manage this usecase without breaking this feature for standard IntelliJ plugin (ex : Java plugin which doesn't use LSP).

**Comment by angelozerr** [comment]
@ericdallo I have started the 0.8.0 release of LSP4IJ which provides type + call hierarchy. Here a sample with clojure:

![image](https://github.com/user-attachments/assets/6a9b1882-ba9b-4e37-ad77-cb54b61a71bd)

See https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md#call-hierarchy to know how to integrate call hierarchy with clojure since you are using a custom clojure language

**Comment by ericdallo** [comment]
@angelozerr thanks for the followup, I will spend more time on this project from now on, and I will take a closer look on lsp4ij, the first thing I need to know is check the features parity, if we can provide all current features the migration seems to worth it!

**Comment by ericdallo** [comment]
@angelozerr I created a branch here to start this integration, some points I noticed:
- clojure-lsp supports intellij 2021 while the plugin supports only 2023 onwards, that's not a big deal though as I plan to support only 2023 onwards as well I guess
- Not sure why the server is not being created automatically even associating with a language
![Image](https://github.com/user-attachments/assets/519646e9-040d-4274-91db-3e912c69983e)
Just wanna mention that it's really important that the clojure-lsp plugin require minimal configs from users as possible, just like it's ATM (where it downloads the server automatically by default and starts it as soon a clojure file is opened).
- After creating manually the server, during the initialize request, I'm not seeing the $progress notifications that server produces during the initialize request, like I mentioned [here](https://github.com/clojure-lsp/clojure-lsp-intellij/issues/53#issuecomment-2199963521), is it required to do any extra config?

- I wanna keep the status bar clojure icon that exists atm, where user has options to see if the LSP is connected without any action and options to restart the server, I suppose I can do that with `LanguageServerManager` class, will take a look later.

**Comment by angelozerr** [comment]
> @angelozerr I created a branch [here](https://github.com/clojure-lsp/clojure-lsp-intellij/pull/60) to start this integration, some points I noticed:
> - clojure-lsp supports intellij 2021 while the plugin supports only 2023 onwards, that's not a big deal though as I plan to support only 2023 onwards as well I guess

Lsp4ij supports 2022.3 to eap (for the moment)

> - Not sure why the server is not being created automatically even associating with a language
> ![Image](https://github.com/user-attachments/assets/519646e9-040d-4274-91db-3e912c69983e)
> Just wanna mention that it's really important that the clojure-lsp plugin require minimal configs from users as possible, just like it's ATM (where it downloads the server automatically by default and starts it as soon a clojure file is opened).

It should be created, that it is strange?

If you restart IJ you dont see your ls?

> - After creating manually the server, during the initialize request, I'm not seeing the $progress notifications that server produces during the initialize request, like I mentioned [here](https://github.com/clojure-lsp/clojure-lsp-intellij/issues/53#issuecomment-2199963521), is it required to do any extra config?

No it should work. Need some investigation 
> 
> - I wanna keep the status bar clojure icon that exists atm, where user has options to see if the LSP is connected without any action and options to restart the server, I suppose I can do that with `LanguageServerManager` class, will take a look later.
> 

The lsp console shows you the server status when it is starting, started, stopping stopped. You can see also lsp traces.

**Comment by ericdallo** [comment]
> Lsp4ij supports 2022.3 to eap (for the moment)

I'm seeing 2023.2 [here](https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions)

> If you restart IJ you dont see your ls?

Same problem after restarting :/ any other guess? this is a important step to integrate

> The lsp console shows you the server status when it is starting, started, stopping stopped. You can see also lsp traces.

I know, but I found that window too hidden, would be nice to keep current behavior.
When server starting or stopped/disabled:

![Image](https://github.com/user-attachments/assets/aa925beb-5bde-4db7-ad8b-46262b6a7be0)

When started and running:
![Image](https://github.com/user-attachments/assets/55a2998c-2fdb-41f0-a2cd-84c52f3ee8e9)

**Comment by angelozerr** [comment]
> > Lsp4ij supports 2022.3 to eap (for the moment)
> 
> I'm seeing 2023.2

Ooops sorry you are rigjt.
 [here](https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions)
> 
> > If you restart IJ you dont see your ls?
> 
> Same problem after restarting :/ any other guess? this is a important step to integrate

If you close the language server view and reopen it?
> 
> > The lsp console shows you the server status when it is starting, started, stopping stopped. You can see also lsp traces.
> 
> I know, but I found that window too hidden, would be nice to keep current behavior.
> When server starting or stopped/disabled:
> 
> ![Image](https://github.com/user-attachments/assets/aa925beb-5bde-4db7-ad8b-46262b6a7be0)
> 
> When started and running:
> ![Image](https://github.com/user-attachments/assets/55a2998c-2fdb-41f0-a2cd-84c52f3ee8e9)

**Comment by ericdallo** [comment]
> If you close the language server view and reopen it?

Nothing, tried multiple things, it's like it's not finding that config in my plugin XML

**Comment by angelozerr** [comment]
> > If you close the language server view and reopen it?
> 
> Nothing, tried multiple things, it's like it's not finding that config in my plugin XML

Indeed it is very strange.

**Comment by angelozerr** [comment]
> > > If you close the language server view and reopen it?
> > 
> > Nothing, tried multiple things, it's like it's not finding that config in my plugin XML
> 
> Indeed it is very strange.

Have you some logs?

**Comment by angelozerr** [comment]
@ericdallo please try my suggestion https://github.com/clojure-lsp/clojure-lsp-intellij/pull/60/files#r1921068929

**Comment by ericdallo** [comment]
@angelozerr that fixes the problem! sorry for my silly fault 😅 

I'll work on the statusbar now, so the only missing important thing for now is the progress notifications not being sent during initialize request, I believe it must be something with those notifications being sent before initialize finish.

I still need to test each feature individually but I believe they should work

**Comment by angelozerr** [comment]
> [@angelozerr](https://github.com/angelozerr) that fixes the problem! sorry for my silly fault 😅

Great!

> 
> I'll work on the statusbar now, so the only missing important thing for now is the progress notifications not being sent during initialize request, I believe it must be something with those notifications being sent before initialize finish.
> 
You mean $progress from LSP or other progress?

> I still need to test each feature individually but I believe they should work

As you are using custom PsiFile, you will have to declare some LSP features like:

 * https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md#signature-help
 * https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md#folding-range
 * https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md#document-symbol
 * https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md#call-hierarchy
 * https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md#type-hierarchy

**Comment by ericdallo** [comment]
> You mean $progress from LSP or other progress kind?

Yes, those are sent during initialize in server side.

**Comment by angelozerr** [comment]
> > You mean $progress from LSP or other progress kind?
> 
> Yes, those are sent during initialize in server side.

Do you see some $progress in the lsp console?

**Comment by ericdallo** [comment]
> Do you see some $progress in the lsp console?

It's weird, but although server is working, the logs and traces are empty

![Image](https://github.com/user-attachments/assets/76c534a6-4b68-48cc-9d3c-19217e5a9ba9)

**Comment by ericdallo** [comment]
@angelozerr what's your suggestion to support auto server download before starting it?
Currently, we have this logic [here](https://github.com/clojure-lsp/clojure-lsp-intellij/blob/master/src/main/clojure/com/github/clojure_lsp/intellij/server.clj#L144-L175) to download server and start it automatically

**Comment by angelozerr** [comment]
> > Do you see some $progress in the lsp console?
> 
> It's weird, but although server is working, the logs and traces are empty
> 
> ![Image](https://github.com/user-attachments/assets/76c534a6-4b68-48cc-9d3c-19217e5a9ba9)

Have you set server trace as verbose?

You can set it inside lsp console. See https://github.com/redhat-developer/lsp4ij/blob/main/docs%2FUserGuide.md

**Comment by ericdallo** [comment]
Ah didn't notice those settings, now I got the trace logs and noticed progress are not being sent by server, probably because server expects a work-done-token sent by client when initializing [here](https://github.com/clojure-lsp/clojure-lsp/blob/1aa48457cb97cd26dca3a8dcc0772fc2e255d985/lib/src/clojure_lsp/startup.clj#L66), I will take a look how to set that token

**Comment by ericdallo** [comment]
@angelozerr I'm not seeing any option to set a work done progress token in the [initParams](https://github.com/redhat-developer/lsp4ij/blob/3cfe729c4590374699cb8cb8ec47b4fc0417b7c3/src/main/java/com/redhat/devtools/lsp4ij/LanguageServerWrapper.java#L302), is that supported?

**Comment by angelozerr** [comment]
> [@angelozerr](https://github.com/angelozerr) I'm not seeing any option to set a work done progress token in the [initParams](https://github.com/redhat-developer/lsp4ij/blob/3cfe729c4590374699cb8cb8ec47b4fc0417b7c3/src/main/java/com/redhat/devtools/lsp4ij/LanguageServerWrapper.java#L302), is that supported?

Unfortunately no. Please create an issue. Is it a blocker issue for you?

**Comment by ericdallo** [comment]
@angelozerr it is as affects feedback for users, and clojure-lsp usually take 10-40s to start depending on the project, that feedback is really valuable.
I can create a issue

**Comment by angelozerr** [comment]
> [@angelozerr](https://github.com/angelozerr) it is as affects feedback for users, and clojure-lsp usually take 10-40s to start depending on the project, that feedback is really valuable. I can create a issue

I understand, please create an issue.
> 
> [@angelozerr](https://github.com/angelozerr) ping about [this point](https://github.com/clojure-lsp/clojure-lsp-intellij/issues/53#issuecomment-2599777306) 😅

I don't know. I will try to do that in the start of the stream connection provider but I don't know if it will impact some performances.

Please try that and if it doesn't work,LSP4IJ could provide a download mechanism.

**Comment by ericdallo** [comment]
[Issue created](https://github.com/redhat-developer/lsp4ij/issues/750).

Will try this in stream connection provider and let you know, thanks!

**Comment by ericdallo** [comment]
@angelozerr I tested some ways but it didn't work with the Background task for some reason:

```clojure
(defn run-background-task! [project title run-fn]
  (let [p (promise)]
    (.run (ProgressManager/getInstance)
          (proxy+
           [project title]
           Task$Backgroundable
            (run [_ ^ProgressIndicator indicator]
              (run-fn indicator))))
    p))

(defn -createConnectionProvider [_ ^Project project]
  (doto (proxy
         [OSProcessStreamConnectionProvider]
         []
          (start []
            @(run-background-task!
              project
              "Downloading clojure-lsp..."
              (fn [_indicator]
                (Thread/sleep 5000) ;; simulate download
                (proxy-super start)))))
    (.setCommandLine (GeneralCommandLine. ["/home/greg/dev/clojure-lsp/clojure-lsp" "listen"]))))
```

I think the best would be a method called before `createConnectionProvider` to prepare/download the server, something like `prepareConnection` or so. WDYT?
Unfortunately, this is a bigger blocker than the other issue

**Comment by angelozerr** [comment]
I agree please create an issue.

**Comment by ericdallo** [comment]
[Done](https://github.com/redhat-developer/lsp4ij/issues/751), thanks for the support @angelozerr

**Comment by angelozerr** [comment]
> [Done](https://github.com/redhat-developer/lsp4ij/issues/751), thanks for the support [@angelozerr](https://github.com/angelozerr)

You are welcome! I need to merge my DAP support (today or tomorrow) and after that I will investigate your issues.

**Comment by ericdallo** [comment]
@angelozerr another possible bug I found:

When finding definition of external libs, clojure-lsp return a jar:file://... file, which works in lsp4ij, opening the file properly, but if closing and opening again it doesn't work, this happens because the didOpen sent by the plugin removes the `jar:` in the URI which looks wrong, so clojure-lsp thinks it's another file.

```javascript
[Trace - 13:55:24] Sending request 'textDocument/definition - (18)'.
Params: {
  "textDocument": {
    "uri": "file:///home/greg/dev/nu/trabalha/src/br/trabalha/br/api/models/transfer_out_request.clj"
  },
  "position": {
    "line": 10,
    "character": 15
  }
}
[Trace - 13:55:26] Received response 'textDocument/definition - (18)' in 2342ms.
Result: [
  {
    "uri": "jar:file:///home/greg/.m2/repository/common-core/common-core/16.71.0/common-core-16.71.0.jar!/common_core/schema.cljc",
    "range": {
      "start": {
        "line": 636,
        "character": 6
      },
      "end": {
        "line": 636,
        "character": 13
      }
    }
  }
]
Params: {
  "textDocument": {
    "uri": "file:///home/greg/.m2/repository/common-core/common-core/16.71.0/common-core-16.71.0.jar!/common_core/schema.cljc",
    "languageId": "clojure",
    "version": 1,
    "text": "..."
  }
}

[Trace - 13:56:07] Sending request 'textDocument/definition - (22)'.
Params: {
  "textDocument": {
    "uri": "file:///home/greg/dev/nu/trabalha/src/br/trabalha/br/api/models/transfer_out_request.clj"
  },
  "position": {
    "line": 10,
    "character": 13
  }
}

[Trace - 13:56:07] Received response 'textDocument/definition - (22)' in 3ms.
Result: [
  {
    "uri": "file:///home/greg/.m2/repository/common-core/common-core/16.71.0/common-core-16.71.0.jar!/common_core/schema.cljc",
    "range": {
      "start": {
        "line": 636,
        "character": 6
      },
      "end": {
        "line": 636,
        "character": 13
      }
    }
  }
]
```

Notice how second definition response from server doesn't have jar: because there was a didOpen without jar: before.
I could add a quick fix in server side, but it doesn't look right to me

**Comment by angelozerr** [comment]
> [@angelozerr](https://github.com/angelozerr) another possible bug I found:
> 
> When finding definition of external libs, clojure-lsp return a jar:file://... file, which works in lsp4ij, opening the file properly, but if closing and opening again it doesn't work, this happens because the didOpen sent by the plugin removes the `jar:` in the URI which looks wrong, so clojure-lsp thinks it's another file.
> 
> [Trace - 13:55:24] Sending request 'textDocument/definition - (18)'.
> Params: {
>   "textDocument": {
>     "uri": "file:///home/greg/dev/nu/trabalha/src/br/trabalha/br/api/models/transfer_out_request.clj"
>   },
>   "position": {
>     "line": 10,
>     "character": 15
>   }
> }
> [Trace - 13:55:26] Received response 'textDocument/definition - (18)' in 2342ms.
> Result: [
>   {
>     "uri": "jar:file:///home/greg/.m2/repository/common-core/common-core/16.71.0/common-core-16.71.0.jar!/common_core/schema.cljc",
>     "range": {
>       "start": {
>         "line": 636,
>         "character": 6
>       },
>       "end": {
>         "line": 636,
>         "character": 13
>       }
>     }
>   }
> ]
> Params: {
>   "textDocument": {
>     "uri": "file:///home/greg/.m2/repository/common-core/common-core/16.71.0/common-core-16.71.0.jar!/common_core/schema.cljc",
>     "languageId": "clojure",
>     "version": 1,
>     "text": "..."
>   }
> }
> 
> [Trace - 13:56:07] Sending request 'textDocument/definition - (22)'.
> Params: {
>   "textDocument": {
>     "uri": "file:///home/greg/dev/nu/trabalha/src/br/trabalha/br/api/models/transfer_out_request.clj"
>   },
>   "position": {
>     "line": 10,
>     "character": 13
>   }
> }
> 
> [Trace - 13:56:07] Received response 'textDocument/definition - (22)' in 3ms.
> Result: [
>   {
>     "uri": "file:///home/greg/.m2/repository/common-core/common-core/16.71.0/common-core-16.71.0.jar!/common_core/schema.cljc",
>     "range": {
>       "start": {
>         "line": 636,
>         "character": 6
>       },
>       "end": {
>         "line": 636,
>         "character": 13
>       }
>     }
>   }
> ]
> Notice how second definition response from server doesn't have jar: because there was a didOpen without jar: before. I could add a quick fix in server side, but it doesn't look right to me

Please create an issue with all bugs you are founding.

**Comment by ericdallo** [comment]
Thanks, [done](https://github.com/clojure-lsp/clojure-lsp-intellij/issues/53)

**Comment by ericdallo** [comment]
@angelozerr thank you for the help so far, I think all known issues are fixed, I will work today in:
- support custom LSP requests that clojure-lsp has
- update docs 
- test and ask for some coworkers to test, Nubank has thousands of devs using Clojure daily so we need to take care as this is a major release.

Besides that, I noticed a small thing: is there any way to increase the diagnostics feedback speed? I'm comparing with Emacs (the editor I use) and the diagnostics in intellij take almost 1 second in increase compared with other editors?

**Comment by angelozerr** [comment]
> [@angelozerr](https://github.com/angelozerr) thank you for the help so far, I think all known issues are fixed, I will work today in:
> 
> * support custom LSP requests that clojure-lsp has
> * double check semantic tokens config
> * update docs
> * test and ask for some coworkers to test, Nubank has thousands of devs using Clojure daily so we need to take care as this is a major release.

Great!

> 
> Besides that, I noticed a small thing: is there any way to increase the diagnostics feedback speed? I'm comparing with Emacs (the editor I use) and the diagnostics in intellij take almost 1 second in increase compared with other editors?

Have you the same problem with your current plugin (without LPP4IJ)?

**Comment by ericdallo** [comment]
> Have you the same problem with your current plugin (without LPP4IJ)?

Yeah, I think so, I feel like it's intellij taking time to call the anonator or do the requests

**Comment by ericdallo** [comment]
@angelozerr when do you intent to do a release? because my co workers need to install the nightly to test, and that's not that convenient

**Comment by ericdallo** [comment]
@angelozerr most clojure-lsp commands return a workspace-edit that should be applied, I think lsp4ij doesn't apply that automatically or I don't know what should I call to apply it after getting the response
```clojure
(defn execute-command [^String name ^String text ^Project project]
  (-> (CommandExecutor/executeCommand
       (doto (LSPCommandContext. (Command. text name) project)
         (.setPreferredLanguageServerId "clojure-lsp")))
      (.response)))
```

**Comment by angelozerr** [comment]
> @angelozerr when do you intent to do a release? because my co workers need to install the nightly to test, and that's not that convenient

I need to polish our dap support and need to merge PRs from external contributor.

I hope I will finish that this week and create a release next week. But I cannot promise you.

**Comment by angelozerr** [comment]
> @angelozerr found a problem, it seems that no LSP feature works for jar:file uris, any guess why? like if I go to definition of a external jar, I can navigate in that file anymore

Even by overriding uri mechanism?

**Comment by angelozerr** [comment]
> @angelozerr most clojure-lsp commands return a workspace-edit that should be applied, I think lsp4ij doesn't apply that automatically or I don't know what should I call to apply it after getting the response
> ```clojure
> (defn execute-command [^String name ^String text ^Project project]
>   (-> (CommandExecutor/executeCommand
>        (doto (LSPCommandContext. (Command. text name) project)
>          (.setPreferredLanguageServerId "clojure-lsp")))
>       (.response)))
> ```

If I remember it should apply edit if response is a textedit or workspace edit 

Could you share a lsp trace sample please.

**Comment by ericdallo** [comment]
> Could you share a lsp trace sample please.

I just noticed the request is not even being sent, it might be my bad, it's weird though as the action is being registered dynamically [here](https://github.com/clojure-lsp/clojure-lsp-intellij/pull/60/files#diff-3fb00e34d0754ae92ec7aafed356598fca1ed5708afcbfd9083c8544bc57b21bR74-R76)

**Comment by angelozerr** [comment]
> > Even by overriding uri mechanism?
> 
> Yes, the go to the file works after overwriting the mechanism, but no LSP feature works, like the mapping of the Clojure lang for the LanguageServer is not being activated for read only files

Have you add proper mapping?

**Comment by ericdallo** [comment]
> Have you add proper mapping?



I think [only this](https://github.com/clojure-lsp/clojure-lsp-intellij/blob/0c768942ce2151aeadff2b04f0670fe461473a34/src/main/resources/META-INF/plugin.xml#L21) should be enough right?

**Comment by ericdallo** [comment]
> I just noticed the request is not even being sent, it might be my bad, it's weird though as the action is being registered dynamically [here](https://github.com/clojure-lsp/clojure-lsp-intellij/pull/60/files#diff-3fb00e34d0754ae92ec7aafed356598fca1ed5708afcbfd9083c8544bc57b21bR74-R76)

@angelozerr I believe it's failing on this [return](https://github.com/redhat-developer/lsp4ij/blob/main/src/main/java/com/redhat/devtools/lsp4ij/commands/LSPCommandAction.java#L32), any guess?

**Comment by ericdallo** [comment]
AFAICS the problem is that we register the `LSPCommandAction` but it's triggered by shortcuts or IntelliJ menus, not your plugin, and fall on that `return`, am I right?

**Comment by angelozerr** [comment]
> > I just noticed the request is not even being sent, it might be my bad, it's weird though as the action is being registered dynamically [here](https://github.com/clojure-lsp/clojure-lsp-intellij/pull/60/files#diff-3fb00e34d0754ae92ec7aafed356598fca1ed5708afcbfd9083c8544bc57b21bR74-R76)
> 
> @angelozerr I believe it's failing on this [return](https://github.com/redhat-developer/lsp4ij/blob/main/src/main/java/com/redhat/devtools/lsp4ij/commands/LSPCommandAction.java#L32), any guess?

Have you registered your command in plugin.xml?

**Comment by angelozerr** [comment]
> > Even by overriding uri mechanism?
> 
> Yes, the go to the file works after overwriting the mechanism, but no LSP feature works, like the mapping of the Clojure lang for the LanguageServer is not being activated for read only files

You mean that if file is read only there is no lsp features ? Which lsp features should work? Codelens ? Have you the same problem with your plugin?

**Comment by ericdallo** [comment]
> Have you registered your command in plugin.xml?

no, because I registered dynamically [here](https://github.com/clojure-lsp/clojure-lsp-intellij/blob/lsp4ij/src/main/clojure/com/github/clojure_lsp/intellij/extension/register_actions_startup.clj#L97) because there are so many otherwise we would need to create lots of files for each one, this already worked in master, I belive the problem is that is not possible to call a LSPCommandAction outside the lsp4ij context, for example via a keybinding or menu action

**Comment by angelozerr** [comment]
> > Have you registered your command in plugin.xml?
> 
> no, because I registered dynamically [here](https://github.com/clojure-lsp/clojure-lsp-intellij/blob/lsp4ij/src/main/clojure/com/github/clojure_lsp/intellij/extension/register_actions_startup.clj#L97) because there are so many otherwise we would need to create lots of files for each one, this already worked in master, I belive the problem is that is not possible to call a LSPCommandAction outside the lsp4ij context, for example via a keybinding or menu action

LSPCommandAction is used when language server asks to the client, please  execute me this command. A good sample with clojure lsp is for codelens when yiu click on codelens a registered LSPCommandAction must be executed.

Why you want to execute a command with a menu item? To consume a custom lsp request?

**Comment by ericdallo** [comment]
Because [there are clojure-lsp commands](https://clojure-lsp.io/features/#execute-command) like paredit that user can execute any time via keybindings, it's really important to be able to have them as actions that user can execute anytime

**Comment by angelozerr** [comment]
> Because [there are clojure-lsp commands](https://clojure-lsp.io/features/#execute-command) like paredit that user can execute any time via keybindings, it's really important to be able to have them as actions that user can execute anytime

Ok so I think you want to do that https://github.com/redhat-developer/lsp4ij/blob/main/docs/DeveloperGuide.md#execute-a-command, right?

**Comment by ericdallo** [comment]
@angelozerr Exactly, i coded that [here](https://github.com/clojure-lsp/clojure-lsp-intellij/blob/0c768942ce2151aeadff2b04f0670fe461473a34/src/main/clojure/com/github/clojure_lsp/intellij/client.clj#L14-L18), the problem is that code is not even being called, because the action is not triggering [here](https://github.com/clojure-lsp/clojure-lsp-intellij/blob/0c768942ce2151aeadff2b04f0670fe461473a34/src/main/clojure/com/github/clojure_lsp/intellij/extension/register_actions_startup.clj#L75)

**Comment by angelozerr** [comment]
I am little lost. Could you try to define the command in plugin.xml and see if it is working.

**Comment by ericdallo** [comment]
Not really, it's simple: server supports lots of `workspace/executeCommand` and it returns workspace-edits which client should apply, but I don't think that's the problem here, because the executeCommand is not even requested, the problem I think is that we are registering an action (LSPCommandAction) that when triggered via shortcut or right click menu it does nothing, the `commandPerformed` is not called

**Comment by angelozerr** [comment]
> Not really, it's simple: server supports lots of `workspace/executeCommand` and it returns workspace-edits which client should apply, but I don't think that's the problem here, because the executeCommand is not even requested, the problem I think is that we are registering an action (LSPCommandAction) that when triggered via shortcut or right click menu it does nothing, the `commandPerformed` is not called, maybe I'm misunderstanding something

As I explained you LSPCommandAction is used to define a command that language server requires like clojure codelens click. This action receive the lsp command to execute which comes from the server.

In oherwise LSPCommandAction must not be used to be triggered by shortcut for instance.

Why you dont use standard IJ base action. In this implementation you call just CommandExecutor#,executeCommand

**Comment by angelozerr** [comment]
@ericdallo here a sample code which execute workspace command and try to convert the response to WorkspaceEdit and apply it.

I have not tested the code, but here the idea:

```java
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.project.Project;
import com.redhat.devtools.lsp4ij.JSONUtils;
import com.redhat.devtools.lsp4ij.LSPIJUtils;
import com.redhat.devtools.lsp4ij.commands.CommandExecutor;
import com.redhat.devtools.lsp4ij.commands.LSPCommandContext;
import org.eclipse.lsp4j.Command;
import org.eclipse.lsp4j.WorkspaceEdit;
import org.jetbrains.annotations.NotNull;

public class SomeAction extends AnAction {
    @Override
    public void actionPerformed(@NotNull AnActionEvent e) {
        Project project = e.getProject();
        if (project == null) {
            return;
        }
        
        // Execute "command.from.your.ls" from your language server
        Command command = new Command("My command", "command.from.your.ls");
        LSPCommandContext commandContext = new LSPCommandContext(command, project);
        commandContext.setPreferredLanguageServerId("myLanguageServerId");
        CommandExecutor.executeCommand(commandContext)
                .response()
                .thenAccept(r -> {
                    // Try to convert the execute response to an LSP WorkspaceEdit 
                    WorkspaceEdit edit =  JSONUtils.toModel(r, WorkspaceEdit.class);
                    if (edit != null) {
                        WriteCommandAction.runWriteCommandAction(project, () -> {
                            LSPIJUtils.applyWorkspaceEdit(edit);
                        });
                    }
                });
    }
}

```

**Comment by ericdallo** [comment]
@angelozerr yeah, I think the only way will be to have AnAction for those commands and declare LSPCommandAction for the ones returned by server like the code lens.

I'm trying to follow your snippet but when consuming the `response` future, I get this exception:
```java
Caused by: java.lang.NullPointerException: Cannot invoke "com.redhat.devtools.lsp4ij.LanguageServerItem.getServerDefinition()" because "languageServer" is null
        at com.redhat.devtools.lsp4ij.commands.CommandExecutor.lambda$executeCommandServerSide$1(CommandExecutor.java:183)
        at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:990)
        at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:974)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:209)
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)
        at com.redhat.devtools.lsp4ij.LanguageServerWrapper.lambda$start$4(LanguageServerWrapper.java:315)
```

**Comment by angelozerr** [comment]
> [@angelozerr](https://github.com/angelozerr) yeah, I think the only way will be to have AnAction for those commands and declare LSPCommandAction for the ones returned by server like the code lens.
> 
> I'm trying to follow your snippet but when consuming the `response` future, I get this exception:
> 
> Caused by: java.lang.NullPointerException: Cannot invoke "com.redhat.devtools.lsp4ij.LanguageServerItem.getServerDefinition()" because "languageServer" is null
>         at com.redhat.devtools.lsp4ij.commands.CommandExecutor.lambda$executeCommandServerSide$1(CommandExecutor.java:183)
>         at java.base/java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:990)
>         at java.base/java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:974)
>         at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
>         at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194)
>         at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:209)
>         at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)
>         at com.redhat.devtools.lsp4ij.LanguageServerWrapper.lambda$start$4(LanguageServerWrapper.java:315)
> (defn execute-command [^String name ^String text ^Project project]
>   (-> (CommandExecutor/executeCommand
>         (doto (LSPCommandContext. (Command. text name) project)
>           (.setPreferredLanguageServerId "clojure-lsp")))
>       (.response)
>       deref)) ;; throws exception

Please reinstall nighty build. The NPE error should be fixed and it should show an error coming from the language server (the usecase for you).

**Comment by ericdallo** [comment]
tks, makes sense I forgot to send the args of the command, will do that

**Comment by angelozerr** [comment]
> tks, makes sense I forgot to send the args of the command, will do that 

You are saying that because you have installed nighty and now you see proper error?

**Comment by angelozerr** [comment]
When you will provide those action, is there other things which block you to adopt lsp4ij ?

**Comment by angelozerr** [comment]
> > You are saying that because you have installed nighty and now you see proper error?
> 
> no, I just forgot to send the command args, all clojure-lsp custom commands receive the current cursor location as args, will fix it.
> 
> > When you will provide those actions, is there other things which block you to adopt lsp4ij ?
> 
> there is still the bug of after going to a definition of a external dependency (like a jar), no LSP feature works (code lends, definition, references) in that file, the only request sent after going to that is the `textDocument/definition` request, but the plugin doesn't send any other requests like `definition` or `references`:
> ```
> [Trace - 09:02:44] Sending notification 'textDocument/didOpen'
> Params: {
>   "textDocument": {
>     "uri": "jar:file:///home/greg/.m2/repository/prismatic/schema/1.4.1/schema-1.4.1.jar!/schema/core.cljc",
>     "languageId": "clojure",
>     "version": 1,
>     "text": "..."
>   }
> }
> ```

Have you the same problem with your plugin?

**Comment by ericdallo** [comment]
> Have you the same problem with your plugin?

Yes, since my plugin implements manually the client requests

**Comment by angelozerr** [comment]
> > Have you the same problem with your plugin?
> 
> No, since my plugin implements manually the client requests, we don't have that problem there

Perhaps it is a limitation of IJ which disable those features in ReadOnly mode?

**Comment by ericdallo** [comment]
> Perhaps it is a limitation of IJ which disable those features in ReadOnly mode?

I don't think so, since current plugin master also puts the file as read only.
LMK if you wanna give it a try, although it's clojrue I can give you the step by step to repro the problem, I have a https://github.com/ericdallo/clojure-sample project which you could try, but it's basically finding a definition of a external dep, then you would see no feature works after that 

![Image](https://github.com/user-attachments/assets/a4da3582-f5c9-47c9-93d3-ad3a7ed8c0b6)

**Comment by angelozerr** [comment]
Could you create an issue please with this problem

**Comment by ericdallo** [comment]
[Done](https://github.com/redhat-developer/lsp4ij/issues/782)

**Comment by ericdallo** [comment]
I improved the semantic tokens coloring @angelozerr , just sharing with you the comparison:

![Image](https://github.com/user-attachments/assets/97daf243-cb2f-4917-b6ac-c3480b10ab0e)
![Image](https://github.com/user-attachments/assets/d98a4755-3897-43e4-b8e6-0f57ee7b1452)

**Comment by ericdallo** [comment]
BTW I started to track what's missing and what's done on the description of[ the PR](https://github.com/clojure-lsp/clojure-lsp-intellij/pull/60#issue-2796047928)

**Comment by angelozerr** [comment]
> I improved the semantic tokens coloring [@angelozerr](https://github.com/angelozerr) , just sharing with you the comparison of without semantic-tokens and with them:
> 
> ![Image](https://github.com/user-attachments/assets/97daf243-cb2f-4917-b6ac-c3480b10ab0e) ![Image](https://github.com/user-attachments/assets/d98a4755-3897-43e4-b8e6-0f57ee7b1452)

Nice! Glad LSP4IJ is enough extensible to manage that.

**Comment by ericdallo** [comment]
Overall, lsp4ij API is really good IMO, way better than IntelliJ API, congrats on that

**Comment by angelozerr** [comment]
> Overall, lsp4ij API is really good IMO, way better than IntelliJ API, congrats on that

Thanks so much for this kind of feedback! This API comes from @fbricon

**Comment by ericdallo** [comment]
@angelozerr for some reason after recent changes the commands execution following [this](https://github.com/clojure-lsp/clojure-lsp-intellij/blob/lsp4ij/src/main/clojure/com/github/clojure_lsp/intellij/client.clj#L50) are throwing this exception, any guess?

```java
2025-02-04 10:13:49,621 [1848379]   INFO - STDERR - com.intellij.openapi.progress.ProcessCanceledException: com.intellij.platform.instanceContainer.internal.Con
tainerDisposedException: Container 'ProjectImpl@1181426054 services' was disposed
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.intellij.serviceContainer.ComponentManagerImpl.doGetService(ComponentManagerImpl.kt:722)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.intellij.serviceContainer.ComponentManagerImpl.getService(ComponentManagerImpl.kt:695)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.redhat.devtools.lsp4ij.settings.UserDefinedLanguageServerSettings.getInstance(UserDefinedLanguage
ServerSettings.java:59)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.redhat.devtools.lsp4ij.commands.LSPCommandContext.<init>(LSPCommandContext.java:80)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.redhat.devtools.lsp4ij.commands.LSPCommandContext.<init>(LSPCommandContext.java:72)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.github.clojure_lsp.intellij.client$execute_command.invokeStatic(client.clj:50)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.github.clojure_lsp.intellij.client$execute_command.invoke(client.clj:49)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.github.clojure_lsp.intellij.extension.register_actions_startup$on_action_performed.invokeStatic(r
egister_actions_startup.clj:81)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.github.clojure_lsp.intellij.extension.register_actions_startup$on_action_performed.invoke(registe
r_actions_startup.clj:78)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at clojure.core$partial$fn__5931.invoke(core.clj:2655)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.github.ericdallo.clj4intellij.action$register_action_BANG_$actionPerformed2919_impl__2921.invoke(
action.clj:27)
2025-02-04 10:13:49,622 [1848380]   INFO - STDERR -     at com.github.ericdallo.clj4intellij.action.proxy_plus2918.actionPerformed(Unknown Source)
2025-02-04 10:13:49,623 [1848381]   INFO - STDERR -     at com.intellij.openapi.actionSystem.ex.ActionUtil.doPerformActionOrShowPopup(ActionUtil.kt:374)
2025-02-04 10:13:49,623 [1848381]   INFO - STDERR -     at com.intellij.openapi.keymap.impl.ActionProcessor.performAction(ActionProcessor.java:32)
2025-02-04 10:13:49,623 [1848381]   INFO - STDERR -     at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher$actionProcessor$1.performAction(IdeKeyEventDis
patcher.kt:499)
```

**Comment by ericdallo** [comment]
It's curious that it doesn't happen all the time, restarting intellij seems to make it work again

**Comment by angelozerr** [comment]
No I have no idea why you have this problem?

Have you closed some IJ project?

**Comment by ericdallo** [comment]
Yeah, nvm, it probably was related with testing multiple projects at once, I can't repro anymore, will wait happen again to debug further more.

**Comment by ericdallo** [comment]
Avaiable in 3.0.0 plugin, thanks for all the support @angelozerr!

**Comment by angelozerr** [comment]
> Avaiable in 3.0.0 plugin, thanks for all the support [@angelozerr](https://github.com/angelozerr)!

You are welcome! Hope you will not have some regression since you have adopted LSP4IJ which is a major change on your side.

Don't hesitate to create a PR on LSP4IJ github repo to add your plugin and if you wish, don't hesitate to add a review in marketplace to help us to promote LSP4IJ. Thanks!

**Comment by angelozerr** [comment]
> > Avaiable in 3.0.0 plugin, thanks for all the support [@angelozerr](https://github.com/angelozerr)!
> 
> You are welcome! Hope you will not have some regression since you have adopted LSP4IJ which is a major change on your side.
> 
> Don't hesitate to create a PR on LSP4IJ github repo to add your plugin and if you wish, don't hesitate to add a review in marketplace to help us to promote LSP4IJ. Thanks!

@ericdallo I am sorry to insist but could you create a PR inLSP4IJ to add your plugin and add a review on Jetbrains marketplace. It is super important for us to convince other plugins to be based on LSP4IJ.

It would be nice also to add on the bottom of https://github.com/redhat-developer/lsp4ij/blob/main/docs/user-defined-ls/clojure-lsp.md that it is recommanded to install your plugin which is based on LSP4IJ by given some advantages like "installation server", "LSP command implementation by given a sample with codelens).

Thanks so much to help us to promote LSP4IJ.

**Comment by ericdallo** [comment]
Sure, done [here](https://github.com/redhat-developer/lsp4ij/pull/844)

**Comment by angelozerr** [comment]
Thanks so much @ericdallo !
